python/src/equistore/operations/join.py

```diff
@@ -0,0 +1,209 @@
+    for block in blocks[1:]:
+        _check_blocks(block_ref, block, ["samples", "components", "gradients"], fname)
+        _check_same_gradients_components(block_ref, block, fname)
+
+    properties = _join_labels([block.properties for block in blocks])
+    result_block = TensorBlock(
+        values=_dispatch.hstack([block.values for block in blocks]),
+        samples=block_ref.samples,
+        components=block_ref.components,
+        properties=properties,
+    )
+
+    for parameter, gradient_ref in block_ref.gradients():
+
+        # Different blocks might contain different gradient samples
+        gradient_sample_values = np.unique(
+            np.hstack([block.gradient(parameter).samples for block in blocks])
+        ).tolist()
+        gradient_samples = Labels(
+            names=gradient_ref.samples.names, values=np.array(gradient_sample_values)
+        )
+
+        gradient_data = np.zeros(
+            (len(gradient_samples),)
+            + gradient_ref.data.shape[1:-1]
+            + (len(properties),)
+        )
+
+        i_prop = 0
+        for block in blocks:
+            gradient = block.gradient(parameter)
+
+            # find the correct sample position to put the data
+            for i_grad, sample_grad in enumerate(gradient.samples):
+                # TODO: once C-API is changed replace by:
+                # i_sample = gradient_samples.position(sample_grad)`
+                i_sample = np.where(gradient_samples == sample_grad)[0][0]
+                gradient_data[
+                    i_sample, ..., i_prop : i_prop + len(gradient.properties)
+                ] = gradient.data[i_grad]
+
+            i_prop += len(gradient.properties)
+
+        result_block.add_gradient(
+            parameter, gradient_data, gradient_samples, gradient_ref.components
+        )
+
+    return result_block
+
+
+def _join_blocks_along_samples(blocks: List[TensorBlock]) -> TensorBlock:
+    """Join a sequence of :py:class:`TensorBlock` along their samples."""
+    block_ref = blocks[0]
+
+    fname = "_join_blocks_along_samples"
+    for block in blocks[1:]:
+        _check_blocks(
+            block_ref, block, ["properties", "components", "gradients"], fname
+        )
+        _check_same_gradients_components(block_ref, block, fname)
+
+    result_block = TensorBlock(
+        values=_dispatch.vstack([block.values for block in blocks]),
+        samples=_join_labels([block.samples for block in blocks]),
+        components=block_ref.components,
+        properties=block_ref.properties,
+    )
+
+    for parameter, gradient_ref in block_ref.gradients():
+        gradient_data = [gradient_ref.data]
+        gradient_samples = [np.array(gradient_ref.samples.tolist())]
+
+        for block in blocks[1:]:
+            gradient = block.gradient(parameter)
+            gradient_data.append(gradient.data)
+
+            samples = np.array(gradient.samples.tolist())
+            samples += 1 + gradient_samples[-1][:, 0].max()
```

Comment:
```suggestion
            samples[:, 0] += 1 + gradient_samples[-1][:, 0].max()
```

I'm not sure why you don't need to specify which column you are adding 1+max to?

Reply:
Indeed. Maybe numpy magic...

Reply:
Did you try something with position gradients? With cell gradients it would work since there is a single gradient sample named "sample"

Reply:
I tried position gradients and at least it did not throw an error. But I will do some additional tests today.